Automattic/woocommerce-payments
# Use recommended order currency API for multicurrency subscription renewal | switch | resubscribe

Fixes #4152

## Changes proposed in this Pull Request
[Custom order tables are coming to WooCommerce core.](https://developer.woocommerce.com/2022/01/17/the-plan-for-the-woocommerce-custom-order-table/)

To prepare for that, we need to move away from using low-level APIs to access post meta for orders. 

This PR updates the multicurrency WC Subscriptions integration to use the more robust and clearer [`get_currency()`](https://github.com/woocommerce/woocommerce/blob/dc095300c275311f5ae1aeea863c747ef3f435f0/plugins/woocommerce/includes/abstracts/abstract-wc-order.php#L297-L299) API. Previously the code used `get_post_meta`, which is [not recommended](https://woocommerce.com/document/developing-using-woocommerce-crud-objects/), and could break in future when stores migrate to using custom order tables.

## Testing instructions
This change should not change any behaviour. As I understand it, there's no buggy behaviour here unless the store has migrated to custom tables (which is not yet available/shipped yet). 

For my testing, I added logging to confirm the relevant lines of code were run, and logged the currency code (to ensure it was correct). All the flows below trigger the code when preparing an order for cart/checkout. 

In all tests, the currency code should be the shopper's currency, not the store default.

The test instructions below are all a very simple happy path. Reviewers please suggest and test alternative flows that might be impacted.
##### Prerequisite - set up payments, subscription product, and multi-currency
1. Onboard WooCommerce Payments and install and activate WC Subscriptions. 
2. Publish a subscription product: 
  - Go to WooCommerce admin dashboard, click `Products > Add New` in sidebar.
  - Type a product name e.g. `Coffee Subscription`.
  - Scroll down to `Product data` box and select `Simple subscription` from `Product type` dropdown.
  - Enter a price e.g. `20` in `Subscription price ($)` field.
  - Select a renewal term, e.g. `every` `day`.
3. [Set up multi currency.](https://woocommerce.com/document/payments/currencies/multi-currency-setup/) Needs to be activated in Woo settings and WCPay, and a currency switcher displayed for customer on store front end.
  - In WooCommerce admin dashboard, go to `Payments > Settings`.
  - Click `Multi-currency` tab.
  - Click `Add currencies` and add one or more additional currencies, e.g. Yen, Euro etc.
  - Scroll down to `Store settings`.
  - Ensure `Add a currency switcher to the Storefront theme on breadcrumb section.`.
  - Click `Save changes`.

You should now have a subscription product in your store that shoppers can purchase in their chosen currency.

##### Test multicurrency subscription renewal
1. Complete prerequisite setup above.
1. Purchase a subscription in customer currency (not store default).
1. Go to subscription in admin dashboard.
1. Select action `Create pending renewal order`, click `Update`.
1. As customer, go to `My account`, click `Subscriptions`, scroll down to `Related orders`.
1. Click `Pay`. Confirm correct (customer) currency is displayed in checkout.

##### Test multicurrency subscription switch
1. Complete prerequisite setup above.
1. In `WooCommerce > Settings > Subscriptions`, enable both `Allow switching` options. (See below pic)
3. Set up subscription products that [subscriber can switch between (aka upgrade/downgrade)](https://woocommerce.com/document/subscriptions/switching-guide/). Easiest way is setting up a variable subscription product with 2 or more levels (e.g. `Basic | Pro` attribute). Can also switch between [grouped products](https://woocommerce.com/document/managing-products/#adding-a-grouped-product).
   - Click on `Products > Add new` in WooCommerce admin dashboard sidebar.
   - Type a product name e.g. `Coffee Subscription`.
   - Scroll down to `Product data` box and select `Variable subscription` from `Product type` dropdown.
   - Click `Attributes`.
   - Click `Add` next to `Custom product attribute` dropdown.
   - Type `Level` in `Name:` box.
   - Type `Basic | Regular | Pro` in `Value(s)` box. 
   - Check `Used for variations` checkbox.
   - Click `Save attributes`.
   - Click `Variations`.
   - Select `Create variations from all attributes` from dropdown and click `Go`. Click `OK` in `Are you sure…` modal to continue. Click `OK` in `3 variations added` modal.
   - Click disclosure triangle (on right) in first variation and add a price in `Subscription price ($)` field.
   - Repeat for all variations.
   - Click `Save changes`.
   - Scroll up to top of page and click `Publish` on right to publish the variable subscription product.
4. As shopper, subscribe to one of the variation products.
5. Go to `My account`, click `Subscriptions`.
7. Click switchable subscription to view details/options.
1. Click `Upgrade or Downgrade`. Select a variation to switch to and click `Switch subscription`. Confirm correct (customer) currency is displayed in checkout.

<img width="754" alt="Screen Shot 2022-06-14 at 11 15 26 AM" src="https://user-images.githubusercontent.com/4167300/173462568-b92c1669-c701-4a44-8f5b-f35df8be4c84.png">

##### Test multicurrency resubscribe
1. Complete prerequisite setup above.
1. Purchase a subscription in customer currency (not store default).
1. Go to subscription in admin dashboard.
1. Click `Cancel` for the subscription. Subscription should now be cancelled, so shopper can resubscribe.
1. As customer, go to `My account`, click `Subscriptions`
3. Click `Resubscribe`. Refer to [docs](https://woocommerce.com/document/subscriptions/resubscribe/) for details - resubscribing is a kind of renewal. 
4. Confirm correct (customer) currency is displayed in checkout.

-------------------


**Post merge**



- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : [Test multicurrency subscription renewal, switch and re-subscribe ](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions-for-WC-Payments-4.4.0#test-multicurrency-subscription-renewal-switch-and-re-subscribe-related-pull-request)
